driver/src/main/java/org/neo4j/driver/internal/cluster/loadbalancing/LoadBalancer.java

```diff
@@ -44,56 +48,83 @@
 import org.neo4j.driver.v1.exceptions.ServiceUnavailableException;
 import org.neo4j.driver.v1.exceptions.SessionExpiredException;
 
+import static java.util.concurrent.CompletableFuture.completedFuture;
+import static org.neo4j.driver.internal.async.Futures.failedFuture;
+
 public class LoadBalancer implements ConnectionProvider, RoutingErrorHandler, AutoCloseable
 {
     private static final String LOAD_BALANCER_LOG_NAME = "LoadBalancer";
 
     private final ConnectionPool connections;
+    private final AsyncConnectionPool asyncConnectionPool;
     private final RoutingTable routingTable;
     private final Rediscovery rediscovery;
     private final LoadBalancingStrategy loadBalancingStrategy;
     private final Logger log;
 
+    private CompletableFuture<RoutingTable> refreshRoutingTableFuture;
+
     public LoadBalancer( BoltServerAddress initialRouter, RoutingSettings settings, ConnectionPool connections,
-            Clock clock, Logging logging, LoadBalancingStrategy loadBalancingStrategy )
+            AsyncConnectionPool asyncConnectionPool, Clock clock, Logging logging,
+            LoadBalancingStrategy loadBalancingStrategy )
     {
-        this( connections, new ClusterRoutingTable( clock, initialRouter ),
+        this( connections, asyncConnectionPool, new ClusterRoutingTable( clock, initialRouter ),
                 createRediscovery( initialRouter, settings, clock, logging ), loadBalancerLogger( logging ),
                 loadBalancingStrategy );
     }
 
     // Used only in testing
-    public LoadBalancer( ConnectionPool connections, RoutingTable routingTable, Rediscovery rediscovery,
-            Logging logging )
+    public LoadBalancer( ConnectionPool connections, AsyncConnectionPool asyncConnectionPool,
+            RoutingTable routingTable, Rediscovery rediscovery, Logging logging )
     {
-        this( connections, routingTable, rediscovery, loadBalancerLogger( logging ),
-                new LeastConnectedLoadBalancingStrategy( connections, logging ) );
+        this( connections, asyncConnectionPool, routingTable, rediscovery, loadBalancerLogger( logging ),
+                new LeastConnectedLoadBalancingStrategy( connections, asyncConnectionPool, logging ) );
     }
 
-    private LoadBalancer( ConnectionPool connections, RoutingTable routingTable, Rediscovery rediscovery, Logger log,
+    private LoadBalancer( ConnectionPool connections, AsyncConnectionPool asyncConnectionPool,
+            RoutingTable routingTable, Rediscovery rediscovery, Logger log,
             LoadBalancingStrategy loadBalancingStrategy )
     {
         this.connections = connections;
+        this.asyncConnectionPool = asyncConnectionPool;
         this.routingTable = routingTable;
         this.rediscovery = rediscovery;
         this.loadBalancingStrategy = loadBalancingStrategy;
         this.log = log;
 
-        refreshRoutingTable();
+        if ( connections != null )
```

Comment:
does this depend only on `connections`? Or should `asyncConnectionPool` has a say in the decision?